### PR TITLE
[Identity] Merging feature/identity/140 back to master

### PR DESCRIPTION
**Pull request by schaabs**
This PR is to merge the feature branch feature/identity/140 back into master. This feature branch was holding updates released in Azure.Idenitty 1.4.0-beta.1 ahead of the 1.3.0 release. Now that 1.3.0 has been released these changes can be merged back making master 1.4.0-beta.2.

### CODE REVIEW

**Review by chlowell — APPROVED** [review]

File: sdk/identity/Azure.Identity/samples/ClientSideUserAuthentication.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+To use the `PersistentTokenCache` to persist the cache of any credential simply set the `TokenCache` option.
+
+```C# Snippet:Identity_ClientSideUserAuthentication_Persist_TokenCache
+var credential = new InteractiveBrowserCredential(new InteractiveBrowserCredentialOptions { TokenCache = new PersistentTokenCache() });
+```
+
+### Persisting the AuthenticationRecord
+
+The `AuthenticationRecord` which is returned from the `Authenticate` and `AuthenticateAsync`, contains data identifying an authenticated account. It is needed to identify the appropriate entry in the `TokenCache` to silently on subsequent executions. There is no sensitive data in the `AuthenticationRecord` so it can be persisted in a non-protected state.
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/samples/ClientSideUserAuthentication.md:80]: " to silently on subsequent executions" ?

File: sdk/identity/Azure.Identity/src/TokenCache.cs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+                _lastUpdated = DateTime.UtcNow;
+            }
+            finally
+            {
+                _lock.Release();
+            }
+
+            if (Updated != null)
+            {
+                await Updated.Invoke(new TokenCacheUpdatedArgs(this)).ConfigureAwait(false);
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/TokenCache.cs:235]: What if there are multiple registrations? You are not awaiting all of them.

File: sdk/identity/Azure.Identity/src/TokenCache.cs [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+            await client.GetAccountsAsync().ConfigureAwait(false);
+
+            return merged;
+        }
+
+        /// <summary>
+        /// Disposes of the <see cref="TokenCache"/>.
+        /// </summary>
+        /// <param name="disposing">Indicates whether managed resources should be disposed.</param>
+        protected virtual void Dispose(bool disposing)
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/TokenCache.cs:262]: Do we define the Dispose just for the `_lock`? IIRC it doesn't need to be disposed if we are not creating the  WaitHandle.

File: sdk/identity/Azure.Identity/src/TokenCacheUpdatedArgs.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        internal TokenCacheUpdatedArgs(TokenCache cache)
+        {
+            Cache = cache;
+        }
+
+        /// <summary>
+        /// The <see cref="TokenCache"/> instance which was updated.
+        /// </summary>
+        public TokenCache Cache { get; }
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/TokenCacheUpdatedArgs.cs:19]: Won't this always be the instance to which you've subscribed to?

File: sdk/identity/Azure.Identity/src/TokenCache.cs [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+                throw new ObjectDisposedException(nameof(TokenCache));
+            }
+
+            await _lock.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                args.TokenCache.DeserializeMsalV3(_data, true);
+
+                _cacheAccessMap.GetOrCreateValue(args.TokenCache).Update();
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/TokenCache.cs:187]: Can we use `AddOrUpdate` here and avoid having the `CacheTimestamp` class?

File: sdk/identity/Azure.Identity/src/TokenCacheUpdatedArgs.cs (hunk 4 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/TokenCacheUpdatedArgs.cs:19]: Yes but you could have a single handler subscribing to the updated events of multiple caches, (this is actually one of the main use cases)  so we pass back the instance which is firing the event. We could have used the EventHandler "object sender" pattern, but decided to add it as a property to the event args instead to make the signature simpler as well as type safe.

File: sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs [hunk 6]
```diff
@@ -71,11 +72,13 @@ internal SharedTokenCacheCredential(string tenantId, string username, TokenCrede
 
             _username = username;
 
+            _skipTenantValidation = (options as SharedTokenCacheCredentialOptions)?.EnableGuestTenantAuthentication ?? false;
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs:75]: Do we have a test for this new feature?

File: sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs [hunk 7]
```diff
@@ -71,11 +72,13 @@ internal SharedTokenCacheCredential(string tenantId, string username, TokenCrede
...
+
             _record = (options as SharedTokenCacheCredentialOptions)?.AuthenticationRecord;
 
             _pipeline = pipeline ?? CredentialPipeline.GetInstance(options);
 
-            _client = client ?? new MsalPublicClient(_pipeline, tenantId, Constants.DeveloperSignOnClientId, null, (options as ITokenCacheOptions) ?? s_DefaultCacheOptions);
+            _client = client ?? new MsalPublicClient(_pipeline, tenantId, (options as SharedTokenCacheCredentialOptions)?.ClientId ?? Constants.DeveloperSignOnClientId, null, (options as ITokenCacheOptions) ?? s_DefaultCacheOptions);
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs:81]: Do we have a test for this new feature?

File: sdk/identity/Azure.Identity/src/PersistentTokenCache.cs [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        private static readonly string s_msalCacheClientId = Guid.NewGuid().ToString();
+
+        private static AsyncLockWithValue<MsalCacheHelper> cacheHelperLock = new AsyncLockWithValue<MsalCacheHelper>();
+        private static AsyncLockWithValue<MsalCacheHelper> s_ProtectedCacheHelperLock = new AsyncLockWithValue<MsalCacheHelper>();
+        private static AsyncLockWithValue<MsalCacheHelper> s_FallbackCacheHelperLock = new AsyncLockWithValue<MsalCacheHelper>();
+        private readonly bool _allowUnencryptedStorage;
+        private readonly string _name;
+
+        /// <summary>
+        /// Creation.
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/PersistentTokenCache.cs:28]: Nit: comment

File: sdk/identity/Azure.Identity/src/PersistentTokenCache.cs [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        /// </summary>
+        /// <param name="allowUnencryptedStorage"></param>
+        public PersistentTokenCache(bool allowUnencryptedStorage = true)
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/PersistentTokenCache.cs:31]: Is this setting that important to have a ctor overload with it?

File: sdk/identity/Azure.Identity/src/MsalClientBase.cs [hunk 10]
```diff
@@ -40,9 +36,7 @@ protected MsalClientBase(CredentialPipeline pipeline, string tenantId, string cl
 
             ClientId = clientId;
 
-            EnablePersistentCache = cacheOptions?.EnablePersistentCache ?? false;
-
-            AllowUnencryptedCache = cacheOptions?.AllowUnencryptedCache ?? false;
+            TokenCache = cacheOptions?.TokenCache;
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/MsalClientBase.cs:39]: Can we remove the ITokenCacheOptions and change the parameter type to TokenCache?

**Comment by pakrym** [comment]
Needs tests for TokenCache and PersistantTokenCache

File: sdk/identity/Azure.Identity/src/TokenCacheUpdatedArgs.cs (hunk 4 above)
Comment (pakrym) [reply sdk/identity/Azure.Identity/src/TokenCacheUpdatedArgs.cs:19]: You can always capture the instance you are subscribing to though.

cache.Updated += async () DoSomethingWith(cache);

Comment (pakrym) [reply sdk/identity/Azure.Identity/src/TokenCacheUpdatedArgs.cs:19]: But maybe the Args type is worth to have just from the perspective of future proofing.

File: sdk/identity/Azure.Identity/src/TokenCache.cs (hunk 5 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/TokenCache.cs:187]: conditional week table doesn't have `AddOrUpdate`

File: sdk/identity/Azure.Identity/src/TokenCache.cs (hunk 3 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/TokenCache.cs:262]: Do we have control over when the `SemaphoreSlim` will need to create the underlying WaitHandle?

File: sdk/identity/Azure.Identity/src/TokenCache.cs (hunk 5 above)
Comment (pakrym) [reply sdk/identity/Azure.Identity/src/TokenCache.cs:187]: https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2.addorupdate?view=net-5.0 

Ah, it's in NS2.1...

File: sdk/identity/Azure.Identity/src/TokenCache.cs (hunk 3 above)
Comment (pakrym) [reply sdk/identity/Azure.Identity/src/TokenCache.cs:262]: Only when you call `AvailableWaitHandle`

File: sdk/identity/Azure.Identity/src/PersistentTokenCache.cs (hunk 9 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/PersistentTokenCache.cs:31]: I agree we probably should move this to an options class. This is the first draft of this API and given that currently we only had one option I'm guessing I did this for expediency. I'd rather defer changing it here where we're trying to merge in the changes from the already release 1.4.0-beta.1 so we have a commit in master that matches the released API.

File: sdk/identity/Azure.Identity/src/TokenCacheUpdatedArgs.cs (hunk 4 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/TokenCacheUpdatedArgs.cs:19]: Yes it's quite likely that the data we pass to this event will grow. There's other information that we could pass to this event to help optimize serialization, but for the first draft we elected to keep it simple.

File: sdk/identity/Azure.Identity/src/PersistentTokenCache.cs (hunk 8 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/PersistentTokenCache.cs]: fixed

File: sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs (hunk 7 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs]: added.

File: sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs (hunk 6 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs:77]: added.

File: sdk/identity/Azure.Identity/samples/ClientSideUserAuthentication.md (hunk 1 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/samples/ClientSideUserAuthentication.md]: fixed

File: sdk/identity/Azure.Identity/src/TokenCache.cs (hunk 2 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/TokenCache.cs]: fixed

File: sdk/identity/Azure.Identity/src/TokenCache.cs (hunk 5 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/TokenCache.cs:187]: Perhaps I should have said intellisense didn't show me `AddOrUpdate` 😉

File: sdk/identity/Azure.Identity/src/TokenCache.cs (hunk 3 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/TokenCache.cs:265]: Perhaps we can remove the need for this class to be IDisposable. As I called out below I'd rather defer changing it here  so the API is consistent with the released 1.4.0-beta.1 API. I've filed issue https://github.com/Azure/azure-sdk-for-net/issues/17124 to investigate and remove if unneeded.

File: sdk/identity/Azure.Identity/src/MsalClientBase.cs (hunk 10 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/MsalClientBase.cs:39]: Yes it's no longer needed, but I don't want to make such a large refactor prior to merging the feature branch back in. I opened https://github.com/Azure/azure-sdk-for-net/issues/17125 to address this afterwards.

**Review by pakrym — APPROVED** [review]

**Pull request merged**